### PR TITLE
Allow Fastboot to work with Lazy Loading Engines

### PR DESCRIPTION
**Pull request by devotox**
cc: @trentmwillis 

This fixes the merge conflicts from @trentmwillis branch and updates the manifest hook to use lazyLoading.enabled flag

### CODE REVIEW

File: lib/engine-addon.js [hunk 1]
```diff
@@ -738,27 +738,29 @@ module.exports = {
...
-          otherAssets,
-          finalStylesTree,
-          vendorJSImportTree,
-          concatEngineTree
-        ];
-
-        // Separate engines routes from the host if includeRoutesInApplication
-        // is false
-        var separateRoutes = this.lazyLoading.includeRoutesInApplication === false;
-        if (separateRoutes) {
```
Comment (trentmwillis) [inline lib/engine-addon.js]: I believe the tests are failing partially due to the collisions here. Basically, all the old stuff should still be in the addon we just want to add `configEnvironment` from the new part to the `finalMergeTrees`, does that make sense?

File: tests/acceptance/lazy-routeable-engine-test.js [hunk 2]
```diff
@@ -48,11 +48,12 @@ moduleForAcceptance('Acceptance | lazy routable engine', {
 });
 
 function verifyInitialBlogRoute(assert, loadEvents, application) {
-  assert.equal(loadEvents.length, 4, 'loaded 4 assets');
+  assert.equal(loadEvents.length, 5, 'loaded 5 assets');
   assert.deepEqual(loadEvents[0].split(SEPARATORS), [ '', 'engines-dist', 'ember-blog', 'assets', 'engine-vendor.css' ], 'loaded engine vendor css');
   assert.deepEqual(loadEvents[1].split(SEPARATORS), [ '', 'engines-dist', 'ember-blog', 'assets', 'engine-vendor.js' ], 'loaded engine vendor js');
   assert.deepEqual(loadEvents[2].split(SEPARATORS), [ '', 'engines-dist', 'ember-blog', 'assets', 'engine.css' ], 'loaded engine css');
   assert.deepEqual(loadEvents[3].split(SEPARATORS), [ '', 'engines-dist', 'ember-blog', 'assets', 'engine.js' ], 'loaded engine js');
+  assert.deepEqual(loadEvents[4].split(SEPARATORS), [ '', 'engines-dist', 'ember-blog', 'config', 'environment.js' ], 'loaded environment js');
```
Comment (trentmwillis) [inline tests/acceptance/lazy-routeable-engine-test.js]: Hmm, looks like this is loading the new `config/environment.js` file that we use for Node land. This is sub-optimal as it shouldn't be needed in the browser.

Seems like we have a couple options:
1. Put the environment config file elsewhere in the output so it doesn't get included in the asset manifest
2. Add an option to the asset manifest generator to ignore the environment config (would require changes to [`ember-asset-loader`](https://github.com/ember-engines/ember-asset-loader))
3. Add an additional build step to remove the environment config entry from the generated manifest

@rwjblue do you have any preference here? I'm leaning towards option #2.

Comment (devotox) [inline tests/acceptance/lazy-routeable-engine-test.js]: What have we decided about this? How do we want to implement it?

Comment (rwjblue) [inline tests/acceptance/lazy-routeable-engine-test.js]: Yes, agreed the second option above seems like the best path forward to me.

(sorry for the delays here)

**Comment by trentmwillis** [comment]
@devotox looks like there are still some test failures; let me know if you'd like some help bringing this over the line (I know this code base is kind of crazy)

**Comment by devotox** [comment]
@trentmwillis Yes please. The errors seem to be with the fact that there it expects 3 assets but it has 4. I thought the updates with the Ember Asset Loader done by @rwjblue  would be enough but I'm guessing there is something more i have to do to exclude the environment config?

**Comment by trentmwillis** [comment]
@devotox the changes needed in ember-asset-loader haven't been made yet. I just opened a PR for those: https://github.com/ember-engines/ember-asset-loader/pull/58. Once that's merged we can ignore the `config/environment.js` files.

**Comment by trentmwillis** [comment]
I've released ember-asset-loader@0.4.0 which enables us to ignore the environment config files. Will try integrating it into this PR soon.

**Comment by devotox** [comment]
Ive now also noticed that this is an issue causing a problem with the integration

https://github.com/emberjs/ember.js/issues/15615

Best regards,
Devonte Emokpae

Software Solutions Architect
Zen__

"Meeting all your software solutions needs"

M: +44 7447 903 346
W: Stack Overflow
L: LinkedIn
E: me@devonte.io

**Comment by devotox** [comment]
Anything more on the integration of ember-asset-loader?

**Comment by rwjblue** [comment]
FWIW emberjs/ember.js#15615 was fixed and released in ember-source@2.15.2.

**Comment by trentmwillis** [comment]
Hate to bounce to yet another PR, but rebasing this was too difficult. I've cleaned up the commits and opened a new PR at https://github.com/ember-engines/ember-engines/pull/494, which I plan on finishing today (assuming nothing goes horribly wrong). Closing this, but will update when the other PR is merged as I know multiple folks are following this thread.

**Pull request closed without merging**